### PR TITLE
Complex Version of torch.remainder

### LINKED ISSUE #38349: NumPy-like Functionality Request Rollup
PyTorch often receives requests for [NumPy](https://numpy.org/)-like functionality. This issue is a "rollup" of these requests. It will be continue to be updated with additional requests and helpful links for implementing them.

Newer contributors, checkout the [Wiki](https://github.com/pytorch/pytorch/wiki) and read the [Contribution Guide](https://pytorch.org/docs/stable/community/contribution_guide.html) for help getting started. 

If one of these issues requires additional discussion then, to avoid cluttering this rollup, please create a new issue titled “Implementing [Request]”, label it with “module: numpy”, and start the more focused discussion there. 

**Functions (Updated for the PyTorch 1.8 release):**

- [ ] [divmod](https://numpy.org/doc/1.18/reference/generated/numpy.divmod.html?highlight=divmod#numpy.divmod) (reserved for onboarding)
- [x] [frexp](https://numpy.org/doc/stable/reference/generated/numpy.frexp.html#numpy.frexp)
- [x] [dsplit](https://numpy.org/doc/1.18/reference/generated/numpy.dsplit.html?highlight=dsplit#numpy.dsplit), [hsplit](https://numpy.org/doc/1.18/reference/generated/numpy.hsplit.html?highlight=hsplit#numpy.hsplit), and [vsplit](https://numpy.org/doc/1.18/reference/generated/numpy.vsplit.html?highlight=vsplit#numpy.vsplit)
- [ ] [block](https://numpy.org/doc/stable/reference/generated/numpy.block.html?highlight=block#numpy.block)
- [ ] [diag_indices](https://numpy.org/doc/stable/reference/generated/numpy.diag_indices.html?highlight=diag_indices#numpy.diag_indices)
- [x] [take_along_axis](https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html?highlight=take_along_axis#numpy.take_along_axis)
- [x] [histogram](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html?highlight=histogram#numpy.histogram)
- [ ] [histogram_bin_edges](https://numpy.org/doc/stable/reference/generated/numpy.histogram_bin_edges.html#numpy.histogram_bin_edges)
- [x] [isin](https://numpy.org/doc/stable/reference/generated/numpy.isin.html)(https://github.com/pytorch/pytorch/issues/3025) (reserved for bootcamp)
- [x] let clamp accept tensors (https://github.com/pytorch/pytorch/issues/2793)
- [x] [corrcoef](https://numpy.org/doc/stable/reference/generated/numpy.corrcoef.html?highlight=corrcoef#numpy.corrcoef), requested in https://github.com/pytorch/pytorch/issues/1254

**Completed Functions**
- [x] [diff](https://numpy.org/doc/stable/reference/generated/numpy.diff.html?highlight=diff#numpy.diff), implemented by @soulitzer in https://github.com/pytorch/pytorch/pull/50569
- [x] [any](https://numpy.org/doc/stable/reference/generated/numpy.any.html) and [all](https://numpy.org/doc/stable/reference/generated/numpy.all.html) supporting all dtypes, implemented by @kshitij12345 in https://github.com/pytorch/pytorch/pull/44790
- [x] [broadcast_to](https://numpy.org/doc/stable/reference/generated/numpy.broadcast_to.html?highlight=broadcast_to#numpy.broadcast_to), implemented by @RockingJavaBean in https://github.com/pytorch/pytorch/pull/48997
- [x] [xlogy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.xlogy.html), see original issue https://github.com/pytorch/pytorch/issues/22656, implemented by @kshitij12345 in https://github.com/pytorch/pytorch/pull/48777
- [x] [floatpower](https://numpy.org/doc/stable/reference/generated/numpy.float_power.html?highlight=float_power#numpy.float_power), implemented by @Kiyosora in https://github.com/pytorch/pytorch/pull/44937
- [x] [msort](https://numpy.org/doc/stable/reference/generated/numpy.msort.html?highlight=msort#numpy.msort), added by @Kiyosora in https://github.com/pytorch/pytorch/pull/48440
- [x] [ldexp](https://numpy.org/doc/stable/reference/generated/numpy.ldexp.html?highlight=ldexp#numpy.ldexp), added by @ranman in https://github.com/pytorch/pytorch/pull/45370
- [x] [column_stack](https://numpy.org/doc/stable/reference/generated/numpy.column_stack.html?highlight=column_stack) and [row_stack](https://numpy.org/doc/stable/reference/generated/numpy.ma.row_stack.html?highlight=row_stack), implemented by @RockingJavaBean in https://github.com/pytorch/pytorch/pull/46313
- [x] [kron](https://numpy.org/doc/stable/reference/generated/numpy.kron.html?highlight=kron#numpy.kron), implemented by @IvanYashchuk in https://github.com/pytorch/pytorch/pull/45358
- [x] [ravel](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html?highlight=ravel#numpy.ravel), implemented by @ejguan in https://github.com/pytorch/pytorch/pull/46098
- [x] [swapaxes](https://numpy.org/doc/stable/reference/generated/numpy.swapaxes.html?highlight=swapaxes#numpy.swapaxes), implemented by @kshitij12345 in https://github.com/pytorch/pytorch/pull/46041
- [x] [tile](https://numpy.org/doc/stable/reference/generated/numpy.tile.html?highlight=tile#numpy.tile), implemented by @arbfay in https://github.com/pytorch/pytorch/pull/47974
- [x] [copysign](https://numpy.org/doc/stable/reference/generated/numpy.copysign.html?highlight=copysign#numpy.copysign), implemented by @ejguan in https://github.com/pytorch/pytorch/pull/46396
- [x] [maximum](https://numpy.org/doc/1.18/reference/generated/numpy.maximum.html?highlight=maximum#numpy.maximum), [minimum](https://numpy.org/doc/1.18/reference/generated/numpy.minimum.html) implemented by @RockingJavaBean in https://github.com/pytorch/pytorch/pull/42579
- [x]  [dstack](https://numpy.org/doc/1.18/reference/generated/numpy.dstack.html?highlight=dstack), [hstack](https://numpy.org/doc/1.18/reference/generated/numpy.hstack.html?highlight=hstack#numpy.hstack), [vstack](https://numpy.org/doc/1.18/reference/generated/numpy.vstack.html?highlight=vstack#numpy.vstack), implemented by @muthuArivoli in https://github.com/pytorch/pytorch/pull/42799
- [x] [arcosh](https://numpy.org/doc/1.18/reference/generated/numpy.arccosh.html?highlight=arccosh#numpy.arccosh), [arcsinh](https://numpy.org/doc/1.18/reference/generated/numpy.arcsinh.html?highlight=arcsinh#numpy.arcsinh), [arctanh](https://numpy.org/doc/1.18/reference/generated/numpy.arctanh.html?highlight=arctanh#numpy.arctanh), implemented by @krshrimali in https://github.com/pytorch/pytorch/pull/38388
- [x] [atleast_1d](https://numpy.org/doc/1.18/reference/generated/numpy.atleast_1d.html?highlight=atleast_1d#numpy.atleast_1d), [atleast_2d](https://numpy.org/doc/1.18/reference/generated/numpy.atleast_2d.html), [atleast_3d](https://numpy.org/doc/1.18/reference/generated/numpy.atleast_3d.html), implemented by @kshitij12345 in #41317 
- [x] [deg2rad](https://numpy.org/doc/1.18/reference/generated/numpy.deg2rad.html?highlight=deg2rad), [rad2deg](https://numpy.org/doc/1.18/reference/generated/numpy.rad2deg.html?highlight=rad2deg), implemented by @aayn in https://github.com/pytorch/pytorch/pull/38852
- [x] [fliplr](https://numpy.org/doc/1.18/reference/generated/numpy.fliplr.html?highlight=fliplr), [flipud](https://numpy.org/doc/1.18/reference/generated/numpy.flipud.html?highlight=flipud), implemented by @kshitij12345 in https://github.com/pytorch/pytorch/pull/38599
- [x] [isneginf](https://numpy.org/doc/1.18/reference/generated/numpy.isneginf.html?highlight=isneginf#numpy.isneginf), [isposinf](https://numpy.org/doc/1.18/reference/generated/numpy.isposinf.html?highlight=isposinf#numpy.isposinf), implemented by @RockingJavaBean in #41588
- [x] [isreal](https://numpy.org/doc/1.18/reference/generated/numpy.isreal.html?highlight=isreal#numpy.isreal), implemented by @Justin-Huber in #41298
- [x] [logaddexp](https://numpy.org/doc/1.18/reference/generated/numpy.logaddexp.html?highlight=logaddexp#numpy.logaddexp), [logaddexp2](https://numpy.org/doc/1.18/reference/generated/numpy.logaddexp2.html?highlight=logaddexp#numpy.logaddexp2), implemented by @cloudhan in #38384 
- [x] [signbit](https://numpy.org/doc/1.18/reference/generated/numpy.signbit.html?highlight=signbit#numpy.signbit), implemented by @Kiyosora in #41589
- [x] [count_nonzero](https://numpy.org/doc/1.18/reference/generated/numpy.count_nonzero.html?highlight=count_nonzero#numpy.count_nonzero), implemented by @kshitij12345 in #39992 
- [x] [gcd](https://numpy.org/doc/1.18/reference/generated/numpy.gcd.html?highlight=gcd#numpy.gcd), [lcm](https://numpy.org/doc/1.18/reference/generated/numpy.lcm.html?highlight=lcm#numpy.lcm), implemented by @aayn in #40651
- [x] [moveaxis](https://numpy.org/doc/1.18/reference/generated/numpy.moveaxis.html?highlight=moveaxis) (https://github.com/pytorch/pytorch/issues/36048), implemented by @kshitij12345 in #41480 (as movedim)
- [x] [quantile](https://numpy.org/doc/1.18/reference/generated/numpy.quantile.html?highlight=quantile#numpy.quantile), implemented by @heitorschueroff in #39417
- [x] [nansum](https://numpy.org/doc/stable/reference/generated/numpy.nansum.html?highlight=nansum#numpy.nansum), implemented by @kshitij12345 in https://github.com/pytorch/pytorch/pull/38628
- [x] [hypot](https://numpy.org/doc/1.18/reference/generated/numpy.hypot.html?highlight=hypot#numpy.hypot) (#22764), implemented by @muthuArivoli in https://github.com/pytorch/pytorch/pull/42291
- [x] [nextafter](https://numpy.org/doc/1.18/reference/generated/numpy.nextafter.html?highlight=nextafter#numpy.nextafter), implemented by @muthuArivoli in https://github.com/pytorch/pytorch/pull/42580

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @mruberry @rgommers @heitorschueroff 
cc @aayn @n-gao

### PR DESCRIPTION
**Pull request by aocsa**
Resolves Complex Version of torch.remainder #38349 and #34266 for CPU/CUDA.

- [ ] Complex Version of torch.remainder
  - [x]  CPU implementation
  - [ ]  Vectorized CPU implementation
  - [x]  CUDA implementation 
  - [ ]  unit tests
  - [ ]  update documentation

### CODE REVIEW

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 1]
```diff
@@ -20,6 +26,18 @@ void remainder_kernel_cuda(TensorIterator& iter) {
...
+  else if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+      gpu_kernel_with_scalars(iter,
+        []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
+          scalar_t mod = a - floor_wrapper (a / b) * b;
+          const scalar_t zero = scalar_t(0); 
+          if ((mod != zero) && ((std::real(b) < std::real(zero)) != (std::real(mod) < std::real(zero)))) 
+            mod += b;
+          return mod;
+        });
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:39]: I am not sure this is the right way to implement `floor` for complex. I would go for
```
floor(z)  = ceil(|z|) * z/|z|.
```

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 2]
```diff
@@ -3,12 +3,18 @@
...
 // NOTE: CUDA on Windows requires that the enclosing function
 // of a __device__ lambda not have internal linkage.
 
 namespace at { namespace native {
 
+template<typename T>
+__host__ __device__ static inline c10::complex<T> floor_wrapper(c10::complex<T> v) {
+  return c10::complex<T>(std::floor(v.real()), std::floor(v.imag()));
+}
+
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:17]: I am not sure this is the right way to implement `floor` for complex. I would go for
```
floor(z)  = ceil(|z|) * z/|z|.
```

Comment (nikitaved) [reply aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:17]: The reason being is that I would assume that `floor(z)` and `z` should have the same Argument (angular information). And taking floor in imaginary and real parts separately kind of breaks the argument.

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 3]
```diff
@@ -20,6 +26,18 @@ void remainder_kernel_cuda(TensorIterator& iter) {
...
     });
+  }
+  else if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+      gpu_kernel_with_scalars(iter,
+        []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
+          scalar_t mod = a - floor_wrapper (a / b) * b;
+          const scalar_t zero = scalar_t(0); 
+          if ((mod != zero) && ((std::real(b) < std::real(zero)) != (std::real(mod) < std::real(zero)))) 
+            mod += b;
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:37]: From the definition of reminder it must be `|mod| < |b|`. So, it is worth testing this as well.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:37]: Also, I am also not sure this is the right way for complex, I am thinking about the following algorithm:
```
r_len = reminder(|a| / |b|)
r_arg = (a - r_len * b) / (a - r_len * b).abs()
return r_len * r_arg
```

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 4]
```diff
@@ -20,6 +26,17 @@ void remainder_kernel_cuda(TensorIterator& iter) {
         return r;
       });
     });
+  }
+  else if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+      gpu_kernel_with_scalars(iter,
+        []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
+          auto q = std::trunc(a.real() / b.real());
+          auto r = std::fmod(a.real(), b.real());
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:35]:
```suggestion
          auto q = std::trunc(a.abs() / b.abs());
          auto r = std::fmod(a.abs(), b.abs()); 
```

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 5]
```diff
@@ -20,6 +26,17 @@ void remainder_kernel_cuda(TensorIterator& iter) {
...
+  }
+  else if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+      gpu_kernel_with_scalars(iter,
+        []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
+          auto q = std::trunc(a.real() / b.real());
+          auto r = std::fmod(a.real(), b.real()); 
+          auto arg = (a - q * b) / std::abs(a - q * b);
+          return arg * r;
+        });
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:38]: It is sufficient to just return `a - q * b`

Comment (aocsa) [reply aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:38]: Thanks @nikitaved. So remainder code will look like this:
```
std::complex<scalar_t> remainder(std::complex<scalar_t> a, std::complex<scalar_t> b) {
    auto q = std::trunc(std::abs(a) / std::abs(b));
    return a - q * b;
}
```

**Comment by aocsa** [comment]
This PR is closed as the original issue https://github.com/pytorch/pytorch/issues/34266 is closed too.

**Pull request closed without merging**